lib/cn/cn_tree.c

```diff
@@ -2241,6 +2241,60 @@ cn_tree_node_mclass(struct cn_tree_node *tn, enum hse_mclass_policy_dtype dtype)
     return mclass_policy_get_type(policy, age, dtype);
 }
 
+void
+cn_tree_node_get_min_key(struct cn_tree_node *tn, void **min_key, uint *min_klen)
+{
+    struct kvset_list_entry *le;
+    void *lock;
+
+    INVARIANT(min_key && min_klen);
+
+    *min_key = NULL;
+    *min_klen = 0;
+
+    rmlock_rlock(&tn->tn_tree->ct_lock, &lock);
+    list_for_each_entry (le, &tn->tn_kvset_list, le_link) {
+        struct kvset *kvset = le->le_kvset;
+        void *key;
+        uint klen;
+
+        kvset_get_min_key(kvset, &key, &klen);
+
+        if (!(*min_key) || keycmp(key, klen, *min_key, *min_klen) < 0) {
+            *min_key = key;
+            *min_klen = klen;
+        }
+    }
+    rmlock_runlock(lock);
+}
+
+void
+cn_tree_node_get_max_key(struct cn_tree_node *tn, void **max_key, uint *max_klen)
+{
+    struct kvset_list_entry *le;
+    void *lock;
+
+    INVARIANT(max_key && max_klen);
+
+    *max_key = NULL;
+    *max_klen = 0;
+
+    rmlock_rlock(&tn->tn_tree->ct_lock, &lock);
+    list_for_each_entry (le, &tn->tn_kvset_list, le_link) {
+        struct kvset *kvset = le->le_kvset;
+        void *key;
+        uint klen;
+
+        kvset_get_max_key(kvset, &key, &klen);
+
+        if (!(*max_key) || keycmp(key, klen, *max_key, *max_klen) > 0) {
+            *max_key = key;
+            *max_klen = klen;
+        }
+    }
+    rmlock_runlock(lock);
```

Comment:
I think we should copy the min key out.  Returning a pointer to a buffer owned by the kvset is not safe -- the kvset can be removed by a spill or other compaction as soon as the tree lock is dropped.

Reply:
Agreed.  Plus the ptr has been deconstified....

Reply:
Agreed, my bad. I didn't realize that the kvset can go away as soon as the tree lock is dropped. Will change this API to accept a buffer and a length parameter and copy out the min/max key...

Reply:
Fixed now.